cmd/promtool/rules.go

```diff
@@ -0,0 +1,208 @@
+			return errs
+		}
+
+		for _, ruleGroup := range rgs.Groups {
+			itv := importer.config.EvalInterval
+			if ruleGroup.Interval != 0 {
+				itv = time.Duration(ruleGroup.Interval)
+			}
+
+			rgRules := make([]rules.Rule, 0, len(ruleGroup.Rules))
+			for _, r := range ruleGroup.Rules {
+				expr, err := importer.groupLoader.Parse(r.Expr.Value)
+				if err != nil {
+					return []error{errors.Wrap(err, filename)}
+				}
+				rgRules = append(rgRules, rules.NewRecordingRule(
+					r.Record.Value,
+					expr,
+					labels.FromMap(r.Labels),
+				))
+			}
+
+			groups[rules.GroupKey(filename, ruleGroup.Name)] = rules.NewGroup(rules.GroupOptions{
+				Name:     ruleGroup.Name,
+				File:     filename,
+				Interval: itv,
+				Rules:    rgRules,
+			})
+		}
+	}
+
+	importer.groups = groups
+	return nil
+}
+
+// ImportAll evaluates all the groups and rules and creates new time series
+// and stores them in new blocks.
+func (importer *RuleImporter) ImportAll(ctx context.Context) []error {
+	var errs = []error{}
+	for _, group := range importer.groups {
+		stimeWithAlignment := group.EvalTimestamp(importer.config.Start.UnixNano())
+
+		for _, r := range group.Rules() {
+			err := importer.ImportRule(ctx, r.Query().String(), stimeWithAlignment, group.Interval())
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+	_, err := importer.writer.Flush()
+	if err != nil {
+		errs = append(errs, err)
+	}
+	return errs
+}
+
+// ImportRule imports the historical data for a single rule.
+func (importer *RuleImporter) ImportRule(ctx context.Context, ruleExpr string, stimeWithAlignment time.Time, internval time.Duration) error {
+	ts := stimeWithAlignment
+	appender := importer.writer.Appender()
+
+	for ts.Before(importer.config.End) {
+		currentBlockEnd := ts.Add(blockSize * time.Hour)
+		if currentBlockEnd.After(importer.config.End) {
+			currentBlockEnd = importer.config.End
+		}
+
+		val, warnings, err := importer.apiClient.QueryRange(ctx,
+			ruleExpr,
+			v1.Range{
+				Start: ts,
+				End:   currentBlockEnd,
+				Step:  importer.config.EvalInterval,
+			},
+		)
+		if err != nil {
+			return err
+		}
+		if warnings != nil {
+			fmt.Fprint(os.Stderr, "warning api.QueryRange:", warnings)
+		}
+
+		var matrix model.Matrix
+		switch val.Type() {
+		case model.ValMatrix:
+			matrix = val.(model.Matrix)
+			for _, sample := range matrix {
+				currentLabels := make(labels.Labels, 0, len(sample.Metric))
+				for k, v := range sample.Metric {
+					currentLabels = append(currentLabels, labels.Label{
+						Name:  string(k),
+						Value: string(v),
+					})
+				}
+				for _, value := range sample.Values {
+					_, err := appender.Add(currentLabels, value.Timestamp.Unix(), float64(value.Value))
+					if err != nil {
+						// todo: handle other errors, i.e. ErrOutOfOrderSample and ErrDuplicateSampleForTimestamp
+						return err
+					}
+				}
+			}
+		default:
+			return errors.New("rule result is wrong type")
+		}
+
+		ts = currentBlockEnd
```

Comment:
can't this result in a time being calculated twice if the alignment is 0?

Reply:
hmm good thought. let me think about that and try to fix.